examples/hosts/app-integration/live-schema-upgrade/README.md

```diff
@@ -0,0 +1,57 @@
+# @fluid-example/app-integration-live-schema-upgrade
+
+This example experiments with an approach for upgrading the schema on a container without disposing it.
+
+Please note that the ideas explored here are experimental and under development. They are not yet recommended for broad use in production.
+
+## Scenario
+
+Once a Fluid container has been created, it will contain some set of persisted data in the form of the summary as well as any unsummarized ops. This persisted data can only be correctly interpreted by a compatible container code (typically the same one that created it, or a newer backwards-compatible one). This container code knows the appropriate data stores to load to interpret the summary and process the outstanding ops, as well as provides public access to those data stores for use.
+
+However, suppose you want to change your application's schema by adding an additional data object, but you don't want to migrate all of the existing container data to a brand new container.
+
+## Strategy overview
+
+There are two main approaches to this problem:
+
+1. Use the latest container schema when creating new containers. For any existing containers, wait for all users to disconnect that were running the previous version. Once all users of the previous schema are disconnected, one of the clients with the latest container schema can upgrade the container to enable the new functionality.
+2. As soon as a user with the new container schema connects to an existing container they immediately upgrade the container and forcefully disconnect (close the container) of any connected clients with the old schema.
```

Comment:
It's hard for me to evaluate these choices, as I think it does not account reality of life. In particular, changes propagate slow through the system, including code changes. This could be because new bits do not atomically show up on all CDN endpoints. It could be because application follows safe velocity principles and makes code changes show up slowly, to small population first, before having enough confidence it's safe to increase exposure (dogfood rings is one of the ways to do so).

The question that is not answered here (and guidance that is not provided here) is what happens when (through these process) client finds itself in a position that it can't understand new schema. The main guidance I think that should be here - all such changes have to be shipped dark, and enabling of conversion should happen only after application developers could confirm that 99.99% of existing sessions out there are already on new code that can read new schema.

And if that's the case, the notion that you need to kick out some "old" seems like should not exist (or rather - it should fall into that 0.01% of sessions).

Reply:
I definitely agree that we need to answer these questions and I'll be taking a look at improving our demo script to address these concerns in the next iteration!